dev/AppLifecycle/AppInstance.cpp

```diff
@@ -333,6 +338,109 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
         return s_current->FindForKey(key.c_str());
     }
 
+    std::wstring AppInstance::GenerateRestartAgentPath()
+    {
+        // Calculate the path to the restart agent as being in the same directory as the current module.
+        wil::unique_hmodule module;
+        THROW_IF_WIN32_BOOL_FALSE(GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, reinterpret_cast<PCWSTR>(AppInstance::RequestRestartNow), &module));
+
+        path modulePath = wil::GetModuleFileNameW<std::wstring>(module.get());
+        return modulePath.parent_path() / c_restartAgentFilename;
+    }
+
+    AppRestartFailureReason AppInstance::RequestRestartNow(hstring const& arguments)
+    {
+        // Report feature usage.
+        static bool featureUsageReported{ false };
+        if (!featureUsageReported)
+        {
+            AppLifecycleTelemetry::RequestRestartNow();
+            featureUsageReported = true;
+        }
+
+        // TODO: Redirect UWP requests to CoreApplication::RequestRestartAsync().
+
+        // For packaged, redirect to the already existing API.
+        /*if (IsPackagedProcess())
+        {
+            return winrt::Windows::ApplicationModel::Core::CoreApplication::RequestRestartAsync(arguments).get();
+        }*/
+
+        // Only one restart can happen at a time.
+        auto mutexName = wil::str_printf<std::wstring>(L"%s_RequestRestartNowInProgress", ComputeAppId().c_str());
+        wil::unique_mutex restartMutex;
+
+        restartMutex.create(mutexName.c_str(), CREATE_MUTEX_INITIAL_OWNER, MUTEX_ALL_ACCESS);
+        if (GetLastError() == ERROR_ALREADY_EXISTS)
```

Comment:
`create` is not documented as having `SetLastError` semantics, so this is not reliable. Instead, I suggest checking the Boolean return from `create`.

Reply:
This code is correct I believe.